devito/operator/operator.py

```diff
@@ -128,100 +127,102 @@ class Operator(Callable):
-        self._includes = list(self._default_includes)
-        self._globals = list(self._default_globals)
-
-        # Required for compilation
-        self._compiler = configuration['compiler']
-        self._lib = None
-        self._cfunction = None
-
-        # References to local or external routines
-        self._func_table = OrderedDict()
-
-        # Internal state. May be used to store information about previous runs,
-        # autotuning reports, etc
-        self._state = self._initialize_state(**kwargs)
-
-        # Form and gather any required implicit expressions
-        expressions = self._add_implicit(expressions)
-
-        # Expression lowering: evaluation of derivatives, flatten vectorial equations,
-        # indexification, substitution rules, specialization
-        expressions = [i.evaluate for i in expressions]
-        expressions = [j for i in expressions for j in i._flatten]
-        expressions = [indexify(i) for i in expressions]
-        expressions = self._apply_substitutions(expressions, subs)
-        expressions = self._specialize_exprs(expressions)
+        # Python-level (i.e., compile time) and C-level (i.e., run time) performance
+        profiler = create_profile('timers')
 
-        # Expression analysis
-        self._input = filter_sorted(flatten(e.reads + e.writes for e in expressions))
-        self._output = filter_sorted(flatten(e.writes for e in expressions))
-        self._dimensions = filter_sorted(flatten(e.dimensions for e in expressions))
+        # Lower input expressions to internal expressions (e.g., attaching metadata)
+        expressions = cls._lower_exprs(expressions, **kwargs)
 
         # Group expressions based on their iteration space and data dependences
         # Several optimizations are applied (fusion, lifting, flop reduction via DSE, ...)
         clusters = clusterize(expressions, dse_mode=set_dse_mode(dse))
-        self._dtype, self._dspace = clusters.meta
 
         # Lower Clusters to a Schedule tree
         stree = st_build(clusters)
 
         # Lower Schedule tree to an Iteration/Expression tree (IET)
         iet = iet_build(stree)
-        iet, self._profiler = self._profile_sections(iet)
-        iet = self._specialize_iet(iet, **kwargs)
 
-        # Derive all Operator parameters based on the IET
+        # Instrument the IET for C-level profiling
+        iet = profiler.instrument(iet)
+
+        # Wrap the IET with a Callable
         parameters = derive_parameters(iet, True)
+        op = Callable(name, iet, 'int', parameters, ())
 
-        # Finalization: introduce declarations, type casts, etc
-        iet = self._finalize(iet, parameters)
+        # Lower IET to a Target-specific IET
+        op, target_state = cls._specialize_iet(op, **kwargs)
 
-        super(Operator, self).__init__(self.name, iet, 'int', parameters, ())
+        # Make it an actual Operator
+        op = Callable.__new__(cls, **op.args)
+        Callable.__init__(op, **op.args)
```

Comment:
Isn't this called at `__new__`?

Reply:
it needs to be called explicitly at this point so that we have a proper signature, which allows us to proceed with the later passes